source/slang/slang-check-constraint.cpp

```diff
@@ -1244,58 +1296,217 @@ bool SemanticsVisitor::TryUnifyTypes(
         return TryUnifyConjunctionType(constraints, unifyCtx, fst, snd);
     }
 
-    // If one of the types is a type pack, we need to recursively unify the element types.
-    if (auto fstTypePack = as<ConcreteTypePack>(fst))
+    // Unwrap ConcreteTypePack and unify types with regard to ExpandType.
+    if ((as<ConcreteTypePack>(fst) && as<ConcreteTypePack>(snd)) ||
+        (as<ConcreteTypePack>(fst) && as<ExpandType>(snd)) ||
+        (as<ExpandType>(fst) && as<ConcreteTypePack>(snd)))
     {
-        if (auto sndTypePack = as<ConcreteTypePack>(snd))
+        // Unwrap ConcreteTypePack to make the logic simpler
+        ShortList<Type*> flattenedFirst;
+        ShortList<Type*> flattenedSecond;
+
+        int fstExpandableCount =
+            UnwrapConcreteTypePackAndCountExpandable(fst, flattenedFirst, constraints.genericDecl);
+        int sndExpandableCount =
+            UnwrapConcreteTypePackAndCountExpandable(snd, flattenedSecond, constraints.genericDecl);
+
+        int fstCount = (int)flattenedFirst.getCount();
+        int sndCount = (int)flattenedSecond.getCount();
+
+        // ExpandType can match to more than one corresponding type.
+        // We need to figure out how many types should be unified per each ExpandType.
+        //
+        // Consider this case:
+        // left = ConcreteTypePack(ExpandType, ExpandType)
+        // right = ConcreteTypePack(int, bool, float, double).
+        //
+        // In this case, we shouldn't be mapping the first ExpandType to int and the second
+        // ExpandType to bool, float, double. Instead, they should evenly divide the second type
+        // pack, so we map first ExpandType with int, bool, and second ExpandType to float, double.
+        //
+        int howManyTypesPerEachExpand = 1;
+        int countDifference = fstCount - sndCount;
```

Comment:
I think this should be `fstExpandableCount - sndExpandableCount`, and always map the side with less expandable to the side with more.

Consider:
```
first = (Expand, Expand)
snd = ()
```
This should still work.